fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/MetaDataProtoEditor.java

```diff
@@ -193,14 +412,196 @@ public static void deprecateRecordType(@Nonnull RecordMetaDataProto.MetaData.Bui
         }
     }
 
+    /**
+     * Rename a record type. This can be used to update any top-level record type defined within the
+     * meta-data's records descriptor, including {@code NESTED} records or the union descriptor. However,
+     * it cannot be used to rename nested messages (i.e., messages defined within other messages) or
+     * records defined in imported files. In addition to updating the file descriptor, if the record type
+     * is not {@code NESTED} or the union descriptor, update any other references to the record type
+     * within the meta-data (such as within index definitions).
+     *
+     * @param metaDataBuilder the meta-data builder
+     * @param recordTypeName the name of the existing top-level record type
+     * @param newRecordTypeName the new name to give to the record type
+     */
+    public static void renameRecordType(@Nonnull RecordMetaDataProto.MetaData.Builder metaDataBuilder,
+                                        @Nonnull String recordTypeName, @Nonnull String newRecordTypeName) {
+        // Create a copy of the records builder (rather than calling metaDataBuilder.getRecordsBuilder()) to avoid
+        // corrupting the records builder in metaDataBuilder before all validation has been done.
+        final DescriptorProtos.FileDescriptorProto records = metaDataBuilder.getRecords();
+        boolean found = false;
+        for (DescriptorProtos.DescriptorProto messageType : records.getMessageTypeList()) {
+            if (messageType.getName().equals(recordTypeName)) {
+                found = true;
+            } else if (messageType.getName().equals(newRecordTypeName)) {
+                throw new MetaDataException("Cannot rename record type to " + newRecordTypeName + " as it already exists");
+            }
+            validateTypeNotInNested(messageType, newRecordTypeName);
```

Comment:
Actually, now that I think about it, I think, this check can be removed.

The idea was to stop someone from taking a schema that looked like this:

```proto
message A {
   message B {
   }
   optional B b = 1;
}
message C {
}
```

And renaming `C` to `B`, thereby making the type that the `b` field in `A` refers to not be correct. However, I think if one has

```proto
message A {
   message B {
   }
   optional B b = 1;
}
message B {
}
```

Then the `b` field would still point to `.A.B` rather than to `.B`, so there wouldn't be any real change. I could verify this and then make this change if desired.

Reply:
I have no objection to including that change, since lots of logic around that is already introduced here.

Reply:
Okay, added in 31dd8f1